katdal/caldata.py

```diff
@@ -0,0 +1,400 @@
+                # interpolate over values, or repeat single value
+                # expected size <ts><ch><pol><ant>
+                if hasattr(self._cal_solns[key]['interp'], '__call__'):
+                    solns = self._cal_solns[key]['interp'](timestamps)
+                else:
+                    solns = np.broadcast_to(self._cal_solns[key]['interp'], timestamps.shape)
+                if key == 'K':
+                    solns = np.exp(solns * self._delay_to_phase)
+                self._cal_solns[key]['solns'] = da.from_array(solns,
+                                                              chunks=(self.ts_chunk_size, self.ch_chunk_size, 1, 1),
+                                                              )
+
+
+def applycal(katdal_obj):
+    """
+    Apply the K, B, G solutions to visibilities at provided timestamps.
+    Optionally recompute the weights as well.
+
+    Returns
+    -------
+    katdal_obj: containing vis and weights (optional) with cal solns applied.
+    """
+
+    katdal_obj.cal_coeffs = []
+    katdal_obj.weight_coeffs = []
+
+    def _cal_calc(katdal_obj):
+        """Calculate calibration and weight coefficients"""
+
+        cal_obj = _cal_setup(katdal_obj)
+        _bls_len = katdal_obj._corrprod_keep.sum()
+        _chan_len = katdal_obj._freq_keep.sum()
+        _time_len = katdal_obj._time_keep.sum()
+
+        if _bls_len != len(cal_obj._cp_lookup):
+            raise CalibrationValueError('Shape mismatch between correlation products.')
+        if _chan_len != len(cal_obj._data_channel_freqs):
+            raise CalibrationValueError('Shape mismatch in frequency axis.')
+        if _time_len != len(katdal_obj.timestamps):
+            raise CalibrationValueError('Shape mismatch in timestamps.')
+
+        # calibrate visibilities
+        K = cal_obj._cal_solns['K']['solns']
+        B = cal_obj._cal_solns['B']['solns']
+        G = cal_obj._cal_solns['G']['solns']
+
+        # ((X*K)/B)/G
+        for idx, cp in enumerate(cal_obj._cp_lookup):
+            scale_coeff = None
+            scale_wght = None
+            # K
+            scale_coeff = K[:, :, cp[0][0], cp[0][1]] * K[:, :, cp[1][0], cp[1][1]].conj()
+            # B
+            scale = B[:, :, cp[0][0], cp[0][1]] * B[:, :, cp[1][0], cp[1][1]].conj()
+            scale_coeff /= scale
+            scale_wght = scale.real**2 + scale.imag**2
+            # G
+            scale = G[:, :, cp[0][0], cp[0][1]] * G[:, :, cp[1][0], cp[1][1]].conj()
+            scale_coeff *= np.reciprocal(scale)
+            scale_wght *= scale.real**2 + scale.imag**2
+            katdal_obj.cal_coeffs.append(scale_coeff)
+            katdal_obj.weight_coeffs.append(scale_wght)
+
+        # coefficient matrices <ts><ch><bl>
+        katdal_obj.cal_coeffs = da.stack(katdal_obj.cal_coeffs, axis=2)
+        katdal_obj.weight_coeffs = da.stack(katdal_obj.weight_coeffs, axis=2)
+        return katdal_obj
+    _cal_calc(katdal_obj)
```

Comment:
It looks like you've just defined this function and then called it here, and don't call it anywhere else. You might as well just inline it.